src/structure.js

```diff
@@ -66,12 +66,13 @@ function Structure (options) {
       Infinity;
   }
 
+   this._referencelisteners = Immutable.Map();
+
   this._pathListeners = Immutable.Map();
   this.on('swap', function (newData, oldData, keyPath) {
-    var listenerData = listMatchingOrCreateEmptyList(self._pathListeners, keyPath);
-    self._pathListeners = listenerData.listeners;
-
-    listenerData.matched.forEach(function (fn) {
+    pathCombinations(keyPath).reduce(function(fns, path) {
+      return fns.concat(self._referencelisteners.get(path, Immutable.Set()));
+    }, Immutable.Set()).forEach(function(fn) {
```

Comment:
Here we potentially iterate twice, and that might be a problem if the `pathCombinations` is a large list and the result of the reduce is as well. Here for instance, you could do this in a single step (or laziness)